choreographer/src/main/java/eu/arrowhead/core/choreographer/database/service/ChoreographerDBService.java

```diff
@@ -619,4 +741,235 @@ public ChoreographerStep getStepById(final long id) {
+
+        if (serviceDefinitionOptional.isPresent()) {
+            for (ChoreographerExecutorServiceDefinitionConnection conn : serviceDefinitionOptional.get().getExecutorConnections()) {
+                responseDTO.getData().add(DTOConverter.convertExecutorToExecutorResponseDTO(conn.getExecutorEntry()));
+            }
+        }
+
+        return responseDTO;
+    }
+
+    public ChoreographerExecutorSearchResponseDTO getExecutorByServiceDefinitionAndMinMaxVersion(final String serviceDefinition, final int minVersion, final int maxVersion) {
+        logger.debug("getExecutorByServiceDefinitionAndMinMaxVersion started...");
+
+        List<ChoreographerExecutor> executorList = new ArrayList<>();
+
+        for (int currentVersion = minVersion; currentVersion <= maxVersion; currentVersion++) {
+            Optional<ChoreographerExecutorServiceDefinition> serviceDefinitionOptional = choreographerExecutorServiceDefinitionRepository
+                    .findByServiceDefinitionNameAndVersion(serviceDefinition, currentVersion);
+
+            if (serviceDefinitionOptional.isPresent()) {
+                for (ChoreographerExecutorServiceDefinitionConnection conn : serviceDefinitionOptional.get().getExecutorConnections()) {
+                    executorList.add(conn.getExecutorEntry());
+                }
+            }
+        }
+
+        return DTOConverter.convertExecutorListToExecutorSearchResponseDTO(executorList);
+    }
+
+    private Page<ChoreographerExecutor> getExecutorEntries(int page, int size, Direction direction, String sortField) {
+        logger.debug("getExecutorEntries started...");
+
+        final int validatedPage = page < 0 ? 0 : page;
+        final int validatedSize = size <= 0 ? Integer.MAX_VALUE : size;
+        final Direction validatedDirection = direction == null ? Direction.ASC : direction;
+        final String validatedSortField = sortField == null ? CoreCommonConstants.COMMON_FIELD_NAME_ID : sortField.trim();
+        if (!ChoreographerExecutor.SORTABLE_FIELDS_BY.contains(validatedSortField)) {
+            throw new InvalidParameterException("Sortable field with reference '" + validatedSortField + "' is not available");
+        }
+
+        try {
+            return choreographerExecutorRepository.findAll(PageRequest.of(validatedPage, validatedSize, validatedDirection, validatedSortField));
+        } catch (final Exception ex) {
+            logger.debug(ex.getMessage(), ex);
+            throw new ArrowheadException(CoreCommonConstants.DATABASE_OPERATION_EXCEPTION_MSG);
+        }
+    }
+
+    public ChoreographerSuitableExecutorResponseDTO getSuitableExecutorIdsByStepId(long stepId) {
+        logger.debug("getSuitableExecutorsByStepId started...");
+
+        try {
+            if (choreographerStepRepository.findById(stepId).isPresent()) {
+                Optional<List<Long>> suitableExecutorIds = choreographerExecutorRepository.findExecutorsByStepId(stepId);
```

Comment:
findExecutorByStepId should be a private method instead of a very difficult query